src/nsls2api/services/facility_service.py

```diff
@@ -137,6 +180,112 @@ async def current_operating_cycle(facility: str) -> Optional[str]:
     return cycle.name
 
 
+@dataclass
+class CycleChangeState:
+    new_cycle: Optional[Cycle] = None
+    previous_cycle: Optional[Cycle] = None
+    is_successful: bool = False
+
+
+@asynccontextmanager
+async def cycle_change_context(facility: str, cycle: str):
+    """
+    Context manager to handle atomic-like operations for changing facility cycles.
+
+    Args:
+        facility (str): The facility name
+        cycle (str): The new cycle name to be set
+
+    Yields:
+        CycleChangeState: Object containing the state of the cycle change operation
+
+    Raises:
+        CycleNotFoundError: If the requested cycle cannot be found
+        CycleUpdateError: If the update operations fail
+        CycleVerificationError: If the final state verification fails
+    """
+    state = CycleChangeState()
+
+    try:
+        # Validate and get the new cycle
+        state.new_cycle = await Cycle.find_one(
+            Cycle.facility == facility,
+            Cycle.name == cycle,
+        )
+        if state.new_cycle is None:
+            raise CycleNotFoundError(facility, cycle)
+
+        # Get the current active cycle
+        state.previous_cycle = await Cycle.find_one(
+            Cycle.facility == facility,
+            Cycle.is_current_operating_cycle == True,  # noqa: E712
+        )
+
+        yield state
+
+        # If we get here without exceptions, mark the operation as successful
+        state.is_successful = True
+
+    except CycleOperationError:
+        state.is_successful = False
+        raise
+
+    except Exception as e:
+        state.is_successful = False
+        raise CycleUpdateError(facility, cycle, str(e))
```

Comment:
Consider using `raise..from` to include the original stack trace in the chain...
```suggestion
        raise CycleUpdateError(facility, cycle, str(e)) from e
```